Ast importer visitors

ganenkokb-yandex opened the pull request:
I've rebased commit from [Evianaive](https://github.com/Evianaive/llvm-project/commits?author=Evianaive) and compiled it.
I hope it will speed up fix for #129393.

Evianaive committed:
Implement missing visit function

ganenkokb-yandex committed:
Fix patch compilation

ganenkokb-yandex committed:
Fix non working code:

 - unchecked errors
 - changes Decls were not mapped with helper GetImportedOrCreateDecl
 - SmallVectors were non initialized

balazske commented:
We should ensure that the code builds with `-DBUILD_SHARED_LIBS=ON` cmake option.  I think this is why I can not build it. I get this error:
```
/usr/bin/ld: tools/clang/lib/AST/CMakeFiles/obj.clangAST.dir/ASTImporter.cpp.o: in function `std::conditional<is_base_of_v<clang::Type, clang::concepts::Requirement>, llvm::Expected<clang::concepts::Requirement const*>, llvm::Expected<clang::concepts::Requirement*> >::type clang::ASTNodeImporter::import<clang::concepts::Requirement>(clang::concepts::Requirement*)':
llvm-project/clang/lib/AST/ASTImporter.cpp:1144: undefined reference to `clang::concepts::TypeRequirement::TypeRequirement(clang::TypeSourceInfo*)'
/usr/bin/ld: llvm-project/clang/lib/AST/ASTImporter.cpp:1207: undefined reference to `clang::concepts::ExprRequirement::ExprRequirement(clang::Expr*, bool, clang::SourceLocation, clang::concepts::ExprRequirement::ReturnTypeRequirement, clang::concepts::ExprRequirement::SatisfactionStatus, clang::ConceptSpecializationExpr*)'
/usr/bin/ld: llvm-project/clang/lib/AST/ASTImporter.cpp:1212: undefined reference to `clang::concepts::ExprRequirement::ExprRequirement(clang::concepts::Requirement::SubstitutionDiagnostic*, bool, clang::SourceLocation, clang::concepts::ExprRequirement::ReturnTypeRequirement)'
/usr/bin/ld: tools/clang/lib/AST/CMakeFiles/obj.clangAST.dir/ASTImporter.cpp.o: in function `void std::_Optional_base_impl<clang::concepts::ExprRequirement::ReturnTypeRequirement, std::_Optional_base<clang::concepts::ExprRequirement::ReturnTypeRequirement, true, true> >::_M_construct<clang::TemplateParameterList*&>(clang::TemplateParameterList*&)':
/usr/include/c++/9/optional:416: undefined reference to `clang::concepts::ExprRequirement::ReturnTypeRequirement::ReturnTypeRequirement(clang::TemplateParameterList*)'
collect2: error: ld returned 1 exit status
```

balazske commented:
I think that `ASTImporter` must be moved into a new component (directory in "clang/lib") to fix this problem. `AST` is already dependency of `Sema` so `Sema` can not be added to `AST` as dependency (but could be added to `ASTImporter` if it would be a new component and `Sema` would not need `ASTImporter`). Or the code of some concept related classes must be moved but this looks more difficult.

balazske commented:
I have now a working patch for "splitting" `ASTImporter` from `AST`. If this change will be accepted it is possible to add the new visit functions. The "split" change may require discussion on discord because it affects clang code layout.

ganenkokb-yandex committed:
Fix isDependent calculation in ReturnTypeRequirement ctor.

ganenkokb-yandex committed:
Add newly imported declarations into imported cache

ganenkokb-yandex commented:
@balazske I've put some more fixes to patch to make it work. And now it is on top of main, and compiles with `-DBUILD_SHARED_LIBS=ON`

Full cmake command is:
`cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld" -DLLVM_TARGETS_TO_BUILD="ARM" -DLLVM_DEFAULT_TARGET_TRIPLE="arm64-apple-darwin24.3.0" -DLLVM_BUILD_TESTS=ON  -DCLANG_BUILD_UNIT_TESTS=ON -DBUILD_SHARED_LIBS=ON ../llvm`

Can you make it a chance once more?

balazske commented on clang/lib/AST/ASTImporter.cpp:
```diff
@@ -1063,6 +1100,146 @@ Expected<LambdaCapture> ASTNodeImporter::import(const LambdaCapture &From) {
+  }
+  }
+}
+
```
This function looks too long and can be split into 3 parts for the different types (case branches).

balazske commented on clang/lib/AST/ASTImporter.cpp:
```diff
@@ -735,6 +740,38 @@ namespace clang {
     // E.g. auto f() { struct X{}; return X(); }
     bool hasReturnTypeDeclaredInside(FunctionDecl *D);
+    
+    Expected<ConstraintSatisfaction> FillConstraintSatisfaction(const ASTConstraintSatisfaction& from) {
```
Variable names should start with uppercase letters. There should be always a space after commas (but clang-format should add it?).

balazske commented on clang/lib/AST/ASTImporter.cpp:
```diff
@@ -7359,6 +7536,100 @@ ExpectedStmt ASTNodeImporter::VisitExpr(Expr *E) {
+    return To;
+  To->setLexicalDeclContext(LexicalDC);
+  LexicalDC->addDeclInternal(To);
+  return Importer.MapImported(D, To);
```
`MapImported` is not needed here. It is only used (from the node import functions) if an existing node is returned by the import function.

balazske commented on clang/lib/AST/ASTImporter.cpp:
```diff
@@ -7379,8 +7650,8 @@ ExpectedStmt ASTNodeImporter::VisitVAArgExpr(VAArgExpr *E) {
-  auto ToSubExpr = importChecked(Err, E->getSubExpr());
-  auto ToWrittenTypeInfo = importChecked(Err, E->getWrittenTypeInfo());
+  auto* ToSubExpr = importChecked(Err, E->getSubExpr());
+  auto* ToWrittenTypeInfo = importChecked(Err, E->getWrittenTypeInfo());
```
These changes look not necessary.

ganenkokb-yandex committed:
Revert "Add newly imported declarations into imported cache"

This reverts commit 3380e263eb06287e8c36ca88c60b1c38ec1ad0ae.

ganenkokb-yandex committed:
Reformat changes.

ganenkokb-yandex commented:
> I would like better if the code would look like in PR #138845.
I've applied your changes merged with some fixes from my PR. Thanks.

> Additionally tests are needed for all of the new `Decl` and `Expr` nodes.
On my way!

ganenkokb-yandex committed:
Code review

ganenkokb-yandex committed:
Appply lost changes in new visit.*decl.

Always use GetImportedOrCreateDecl for these.

ganenkokb-yandex committed:
Add tests for ASTImporter on newly added functions

Test cover all code path including new visit and import
 functions for concept and requirement declarations and
 expressions.

ganenkokb-yandex commented:
@balazske tests are ready

ganenkokb-yandex committed:
Fix runtime error on importing Substitution

 failure in type constraint.

ganenkokb-yandex committed:
Fix crash when lambda is null

balazske reviewed (commented):
Code looks almost acceptable. I could not verify if the tests are sufficient (I am not familiar with this new syntax). Can you test this on large projects that use these types of expressions?

ganenkokb-yandex committed:
Code review

balazske commented:
> Newly added VisitSubstNonTypeTemplateParmPackExpr, VisitPseudoObjectExpr, VisitCXXParenListInitExpr are also without tests.
> I've faced them on my project, and looks like it is not because of newly implemented concepts imports. But I'm not sure. Should I move 'em to separate pull request?

It would be better to have not strictly related changes in a different pull request to make debugging more easy (removal of a commit should not remove unrelated code).

balazske commented:
If `FromTypeRequirement.isSubstitutionFailure()` can be true at compile error it can be still possible to call `testImport` or `getTuDecl` with the code.

balazske reviewed (approved):
I think we can put in this change. I tested it on some projects, the number of crashes got less after this change (it is possible that new ones appeared but the total count was less).

ganenkokb-yandex commented:
> I think we can put in this change. I tested it on some projects, the number of crashes got less after this change (it is possible that new ones appeared but the total count was less).

Thank you for your attention.
What additional steps are expected of me to take the pull request further?

ganenkokb-yandex commented:
@balazske I don't have rights for merging PR into main. Could you please help?
